server/src/gnome/prompter.rs

```diff
@@ -0,0 +1,407 @@
+// org.gnome.keyring.Prompter
+// https://gitlab.gnome.org/GNOME/gcr/-/blob/main/gcr/org.gnome.keyring.Prompter.xml
+
+use clap::error::Result;
+use oo7::dbus::ServiceError;
+use serde::{Deserialize, Serialize};
+use zbus::{
+    interface, proxy,
+    zvariant::{
+        DeserializeDict, NoneValue, Optional, OwnedObjectPath, OwnedValue, SerializeDict, Type,
+        Value,
+    },
+};
+
+use crate::{
+    gnome::secret_exchange::SecretExchange,
+    prompt::{Prompt, PromptRole},
+    service::Service,
+};
+
+#[derive(Debug, DeserializeDict, SerializeDict, Type)]
+#[zvariant(signature = "dict")]
+// GcrPrompt properties: https://gitlab.gnome.org/GNOME/gcr/-/blob/main/gcr/gcr-prompt.c#L95
+// This struct causes "Unexpected non-0 padding byte `1`" error.
+// Use properties: HashMap<&str, OwnedValue> to avoid this.
+pub struct Properties {
+    title: Option<String>,
+    message: Option<String>,
+    description: Option<String>,
+    warning: Option<String>,
+    #[zvariant(rename = "password-new")]
+    password_new: Option<bool>,
+    #[zvariant(rename = "password-strength")]
+    password_strength: Option<u32>,
+    #[zvariant(rename = "choice-label")]
+    choice_label: Option<String>,
+    #[zvariant(rename = "choice-chosen")]
+    choice_chosen: Option<bool>,
+    #[zvariant(rename = "caller-window")]
+    caller_window: Option<String>,
+    #[zvariant(rename = "continue-label")]
+    continue_label: Option<String>,
+    #[zvariant(rename = "cancel-label")]
+    cancel_label: Option<String>,
+}
+
+impl Properties {
+    fn for_lock(object: &str) -> Self {
+        Self {
+            title: None,
+            message: Some(String::from("Lock Keyring")),
+            description: Some(format!("Confirm locking '{}' Keyring", object)),
+            warning: None,
+            password_new: None,
+            password_strength: None,
+            choice_label: None,
+            choice_chosen: None,
+            caller_window: None,
+            continue_label: Some(String::from("Lock")),
+            cancel_label: Some(String::from("Cancel")),
+        }
+    }
+
+    fn for_unlock(object: &str, warning: &str) -> Self {
+        Self {
+            title: Some(String::from("Unlock Keyring")),
+            message: Some(String::from("Authentication required")),
+            description: Some(format!(
+                "An application wants access to the keyring '{}', but it is locked",
+                object
+            )),
+            warning: Some(String::from(warning)),
+            password_new: None,
+            password_strength: None,
+            choice_label: None,
+            choice_chosen: None,
+            caller_window: None,
```

Comment:
The caller window is something you should pass as a parameter as well

Reply:
Done.